### PR TITLE
implement issue-19 add prefer no schedule taint to avoid double draining of pods

### PR DESCRIPTION
**Pull request by damoon**
This implements https://github.com/weaveworks/kured/issues/19

The idea is to set a prefer no schedule taint.
The taint is set when the reboot sentinel is found and removed after the restart.

In case the node can not restart immediately the taint asks the kubernetes scheduler to avoid placing new pods on the tainted node.
But this is only preferred, as a last resort the kubernetes scheduler does place pods onto the node.

### CODE REVIEW

**Comment by damoon** [comment]
The test is failing because the taint code uses the update method, but RBAC does only include patching nodes.

**Comment by damoon** [comment]
it works now :smiley: 
someone wants to review it?

File: cmd/kured/main.go
Comment (evrardjp) [inline cmd/kured/main.go:467]: I am confused as how preferNoSchedule works. 

I thought that the preferNoSchedule flag would be set (3rd argument to true) when:
- The rebootRequired returns true
- The node cannot get a lock because another node has a lock.
Therefore I expected the line 454 to be in here (L467) instead, before the continue. However, if we agree that all nodes should be marked as NoSchedule immediately when we know they need rebooting (during the maintenance window), then it's fine for me. After all, the code is indeed cleaner in the current PR implementation.
As long as we document how we do it (maybe a little comment on L454 would help?).

However, I expected the function to flip the NoSchedule to false at all ticks, not simply at the restart of the daemonset. If the timewindow to trigger all the reboots is too small, some nodes might not have rebooted in the maintenance window, and therefore some ndoes will still marked with no schedule outside the timewindow. That doesn't seem right.

TL;DR: Therefore I would move the L440 inside the block 450-452, move that block above the block of the timewindow (446), move the L454 `preferNoSchedule` in the block L466-468 before the `continue`. But of course, I am easy, and happy to discuss it here.

**Review by evrardjp — CHANGES_REQUESTED** [review]
First, allow me to thank you for this PR! Well thought out.

Please see my comments inline for the first part of the review (maybe a comment, and a little code change?) on the preferNoSchedule code location/behaviour.
Finally, I see a problem with the PR right now (maybe I am mistaken, could you clarify?):
This PR changes the state of the taint `TaintEffectPreferNoSchedule` regardless of not it was tained by kured. This is a problem for me, because one administrator might want to mark a node for NoSchedule for other reasons, and we might remove it without its approval. WDYT?

File: cmd/kured/main.go
Comment (damoon) [reply cmd/kured/main.go:92]: i like it.

File: cmd/kured/main.go
Comment (damoon) [reply cmd/kured/main.go:467]: i assumed the order of the checks would be like this because the coast to execute them.
the window-check is just a calculation, the requires-reboot is a local check and the reboot-blocked check is a rpc.

i did not want to have the remove-taint within the loop, because it could take a long time until a rebooted node get untained.
maybe a additional untaint in the block from 450-453?

The true-false flag was a pain in my eye anyway. What do you think about two functions addPreferNoScheduleTaint and removePreferNoScheduleTaint?

Comment (damoon) [reply cmd/kured/main.go:467]: > This PR changes the state of the taint `TaintEffectPreferNoSchedule` regardless of not it was tained by kured. This is a problem for me, because one administrator might want to mark a node for NoSchedule for other reasons, and we might remove it without its approval. WDYT?

Only the taint "weave.works/kured-node-reboot" (by default) is added or remove by the new behavior. Other taints will be untouched. The code from https://github.com/damoon/kured/blob/19-PreferNoSchedule/cmd/kured/main.go#L326 until https://github.com/damoon/kured/blob/19-PreferNoSchedule/cmd/kured/main.go#L367 ensures only to add/remove this one taint.

**Comment by damoon** [comment]
> This PR changes the state of the taint TaintEffectPreferNoSchedule regardless of not it was tained by kured. This is a problem for me, because one administrator might want to mark a node for NoSchedule for other reasons, and we might remove it without its approval. WDYT?

Only the taint "weave.works/kured-node-reboot" (by default) is added or remove by the new behavior. Other taints will be untouched. The code from https://github.com/damoon/kured/blob/19-PreferNoSchedule/cmd/kured/main.go#L326 until https://github.com/damoon/kured/blob/19-PreferNoSchedule/cmd/kured/main.go#L367 ensures only to add/remove this one taint.

File: cmd/kured/main.go
Comment (evrardjp) [reply cmd/kured/main.go:467]: For addPreferNoScheduleTaint: That's indeed more explicit (especially in this part of the code). But overall it's up to you, you wrote it. It's no big deal for me. However: Let's not copy the function into two (and repeat ourselves):  that sounds even worse :)

I am not sure it's about the cost to execute. Someone with more history of the project shoudl tell :)

I was not fond of having this additional untaint without explanations. For me, all this needs is comments, to explain what the design decisions we made. For example:
`we do this outside the main loop because it's an API task, so it takes more time, for which we want to avoid running in the tick loop. This means that a temporary API issue will cause a crash of the daemonset until its state is correct. This also means that if the maintenance window is too small, the machines can still technically be marked with no schedule outside out the maintenance window`.

**Comment by damoon** [comment]
I was not sure who the code was organized.
Should this get its own package in https://github.com/weaveworks/kured/tree/master/pkg ?

File: cmd/kured/main.go
Comment (damoon) [reply cmd/kured/main.go:376]: I added a comment for the untaint outside the loop,
moved the tainting inside the check for rebootBlocked,
and added a untaint to allow for the the reboot window.

Taints got their own package and better naming, the bool is hidden.

Does this look good to you?

**Review by evrardjp — COMMENTED** [review]
I have not tested it, but I like it.

File: cmd/kured/main.go
Comment (evrardjp) [reply cmd/kured/main.go:376]: :+1:

**Comment by evrardjp** [comment]
CI tested it, so I guess that's a good start.

Should we extend smoke test to ensure that we can schedule a basic workload?

**Comment by damoon** [comment]
I tested the parallel reboot of nodes and the taint creation/removal.
(But this happened before moving the code to its own package.)
I would feel more comfortable, if someone else would check it too.

**Comment by damoon** [comment]
i found an issue:
rebootBlocked checks prometheus for alerts and does not indicate an other node has the lock
i moved adding the taint into the !acquire block.

once the testing is done is push the fix.

**Comment by damoon** [comment]
fixed and tested. works for me.

File: pkg/taints/taints.go
Comment (bboreham) [inline pkg/taints/taints.go:127]: nit: patcht typo

File: pkg/taints/taints.go
Comment (bboreham) [inline pkg/taints/taints.go:43]: `ShouldExist` ?

File: cmd/kured/main.go
Comment (bboreham) [inline cmd/kured/main.go:355]: I'm not entirely happy that this is hitting the API server every time round the loop, when most of the time it will have nothing to do.  

Maybe hold a flag saying whether the taint was previously added?

File: cmd/kured/main.go
Comment (bboreham) [inline cmd/kured/main.go:90]: Maybe allow `""` as an option to disable the feature?

**Review by bboreham — COMMENTED** [review]
I left some minor comments below.

It would be good to mention this feature in the README.

I did not really follow the `test` / `add` patch stuff.  Do you have a reference to some documentation explaining it, or another program that does the same?  (`kubectl taint` does not appear to do it).

Also I think there is another corner case: if `rebootRequired()` goes from true to false the taint will not be removed until the end of the reboot window, or until restart of kured.

File: pkg/taints/taints.go
Comment (damoon) [reply pkg/taints/taints.go:43]: renamed :D

File: pkg/taints/taints.go
Comment (damoon) [reply pkg/taints/taints.go:127]: fixed type

File: cmd/kured/main.go
Comment (damoon) [reply cmd/kured/main.go:90]: done

File: cmd/kured/main.go
Comment (damoon) [reply cmd/kured/main.go:355]: done

**Comment by damoon** [comment]
I updated the parameter section in README.

For patches I found this https://dwmkerr.com/patching-kubernetes-resources-in-golang/ example and for `test` i found http://jsonpatch.com/#test. \
Test is needed to ensure the list of taints did not change on server side while the offset to edit was processed locally. \
I believe kubectl uses update instead of patch and avoids this problem with a optimistic lock. \
This implementation chose patch because patch was already part of the RBAC profile while update node is missing.

I changed the behavior in case `rebootRequired()` switches from true to false to remove the taint.

**Review by bboreham — APPROVED** [review]
Looks good to me.
Thanks for all changes, and for the explanation of why it is done this way.

File: cmd/kured/main.go
Comment (evrardjp) [inline cmd/kured/main.go:365]: if the reboot is Blocked by a pod or prometheus, shouldn't we still unmark the node?
For me it means that this node was voluntarily put out of the loop by an administrator, even if the reboot is required during the maintenance window.

In this case, I think that rebootBlocked should also remove any existing NoSchedule taint.
On the other hand, one could argue that a node "out of the loop" shouldn't be touched _at all_. 

In any case, this decision should be recorded somewhere (in a comment, or in the commit message).

**Review by evrardjp — COMMENTED** [review]
I have added a question/comment to clarify our goals. I would prefer if this is clarified before merging. However, I don't think it's worth blocking this PR from merging if we don't get consensus.

Please note that I didn't test this patch. The code looks okay to me.

File: cmd/kured/main.go
Comment (bboreham) [reply cmd/kured/main.go:365]: Being blocked by prometheus can be from a "pending" alert: in my case I see those come and go quite often.
So I think it would be bad to oscillate the taint; I would say once you need to reboot you shouldn't be scheduling new pods onto that node.  (But it's only "preferred" so may get some if necessary).

Maybe we should make the taint `""` by default, so people can choose to try out the feature and report back?

Comment (evrardjp) [reply cmd/kured/main.go:365]: I am fine with both parts of your answer.

Comment (evrardjp) [reply cmd/kured/main.go:365]: @damoon do you mind updating the taint to "" by default?

Comment (evrardjp) [reply cmd/kured/main.go:365]: Alternatively, we can merge this after 16.0, and I can do another PR with updating the default taint.

Comment (damoon) [reply cmd/kured/main.go:365]: I changed the flag to be disabled by default.

Comment (damoon) [reply cmd/kured/main.go:365]: Thinking about the rebootBlocked() cases.

Situation: A node that waits for a long time to reboot, but is blocked by other rebooting nodes. Then a pod, that blocks reboots, is scheduled onto that node. 
A) The taint is remove. As a result even more work gets scheduled here, and needs to be drained again. This should be avoided.
B) The taint stays. As a result the nodes without the taint get a higher average load, even so this node can not reboot once it is its turn to do so. This should also be avoided.
The same should be true for prometheus alert.

My feeling is, to leave a node, that wants to reboot, but is blocked to do so, tainted. It expresses the intend best.
As mentioned, the taint is "only" preferred, and in case B i trust the kubernetes scheduler and resource requests enough for uneven load not be a real issue.

Comment (evrardjp) [reply cmd/kured/main.go:365]: I agree.

**Review by bboreham — APPROVED** [review]
lgtm, thanks!

**Comment by dholbach** [comment]
Nice work and thanks for hanging in there.

Do you think you could add a note https://github.com/weaveworks/kured/issues/295 about this feature?

**Pull request merged**
